### ISSUE TITLE
Leaf Syntax Proposal

### ISSUE THREAD

**Issue by NathanFlurry** [vapor/vapor#545]
# Feature, Enhancement, or Optimization

# Name of Feature

* Author(s): [Nathan Flurry](https://github.com/NathanFlurry)

## Introduction

Leaf is a great opportunity to create a much better templating and tempting engine.

## Motivation

Leaf should be extensible, readable, and easy to pick up.

### Variables

Variables in Leaf are read from just by type the variable name. Object data can be read from variables by using the dot syntax, just like in Swift. For example, `friend.name` would get the `name` variable from the `friend` object. Variables are named using camel-case.

Variables can also be defined with an `@` prefix. This can be seen in the Blocks section. In addition, variables can be assigned in expressions like `#(@myVar = 5)`.

### Expressions

The expression block simply denotes code for Leaf to interpret. Do use this, simply use the syntax `#(...)` where the content inside the parenthesis is in interpreted code. It could be something like `#(5 + 5)` or `#(friend.name)`. Most Swift operators are supported in Leaf.

### Functions

Functions are used to transform data. For example, to check if two values are equal, one would put `#equal(a, b)` which returns a boolean value. Other examples include `#prefix(friend.name, "Hey, ")`, which would take the friend's name, add `"Hey, "` to it, and return another string. For formatting data, one can use `#format("%d", myVariable)`. Functions can also be overloaded.

### Indexing

Objects and arrays can be indexed just like in Swift. To index an array, simply do `friends[0]`. Similarly, objects can be indexed like `friend["name"]`. This is equivalent to `friend.name`. These values are optional, so it should be checked that the value exists. 

### Embedded Documents

External documents can be embedded by adding `#embed("./other-document.leaf")`. `embed` is just a function, nothing special here.

### Blocks

Blocks can be used to dynamically execute a chunk of code. Leaf provides the blocks `#if` and `#loop`.

If statements are fairly straight forward.

```swift
#if(entering) {
	Hello, there!
} #elseif(leaving) {
	Goodbye!
} #else() {
	I've been here the whole time.
}
```

If statements can also check if a value exists and assign the value to a variable if so.

```swift
#if(friend.job, @job) {
	This friend has a job, #(job).
}
```

Loop statements can be defined like `#loop(variable, start index, end index, increment, inclusive or not)`. For example, this would start at 0 and increment all the way up to 8 (since it's exclusive so it won't execute at 10).

```swift
#loop(@i, 0, 10, 2, false) {
	<p>#(i)</p>
}
```

The increment and inclusivity are assumed to be `1` and `false` if they are excluded. Therefore, `#loop(@i, 0, 10, 1, false) { ... }` is the same as `#loop(@i, 0, 10) { ... }`.

Loops can also be used to iterate through arrays.

```swift
#loop(friends, @friend) {
	<p>I have a friend, #(friend.name).</p>
}
```

The index can be retrieved while iterating an array by doing `#loop(friends, @index, @friend) { ... }`.

The same can be done for objects.

```swift
#loop(friend, @key, @value) {
	<p>Key: #(key), value: #(value)</p>
}
```

### Data types

Leaf supports three types of data:

- String
- Number _(floating point or integer)_
- Bool
- Object
- Array

Strings, numbers, and booleans can be expressed as constants inline the code using their literal value. For example, to check if a variable is equal to a specific constant, one would type `#equals(friend.name, "Bob")` where `"Bob"` is a literal string.

The same applies to numbers by expressing floating point numbers like `3.14` and integers as `42`. Numbers will be rendered with as few decimal points as possible. For example, `2.71000000` will be rendered as `2.71` and `64.0000000` will be rendered as `64`. To avoid this, use the `format` function to format your string using C-style formatting.

#### Leaf Representable Objects

To make an object be able to be represented as a value in Leaf, simply implement the `LeafRepresentable` protocol in your class. If your class already implements `JSONRepresentable`, `LeafRepresentable` will be implemented automatically.

#### Functions

If you want to add a custom function, simply create a class implementing `LeafFunction` and register it with the Leaf provider. For example:

```swift
class MyFunction: LeafFunction {
	let name: String = "myFunction"
	let parameters: [LeafType] = [.any, .string, .number]
	
	static func execute(parameters: [LeafValue]) throws -> LeafValue? {
		return .string("<b>A function was called with parameters \(parameters.join(", "))</b>")
	}
}

drop.leaf.functions += [ MyFunc() ]
```

## Impact

Some concepts proposed would break existing Leaf documents.

## Alternatives considered

Keeping the current syntax.

**Comment by tannernelson** [comment]
The way `#if/else` works right now works on the idea of a tag failing.

For example:

```
#if(stuff.count > 1) {
    bar
} ##uppercase("foo")
```

If the `#if` statement is true, then the body will run otherwise the `#if` will fail. When it fails, it chains to any following tags. In this case, there is one other tag on the chain `#uppercase()`. This cannot fail so it runs every time. In this way `##else` is just an "always true" body that you can stick anywhere.

I'm pretty fond of how the chaining works because it doesn't create special syntax requirements for if/else. We're not trying to recreate Swift here.

```
#if(count > 5) {
    > 5
} ##if(count > 3) {
    > 3 but <= 5
} ##if(true) {
    <= 3
}
```

does the same as

```
#if(count > 5) {
    > 5
} ##if(count > 3) {
    > 3 but <= 5
} ##else {
    <= 3
}
```

I think this failing chain idea is really powerful. It supports if/else without any special logic. (Granted the syntax takes getting used to)

**Comment by NathanFlurry** [comment]
Yeah, I really like that. I didn't realize that was what that was for.

**Comment by tannernelson** [comment]
I definitely think we should add the `#prefix(friend.name, "Hey, ")` and `#format` commands.

**Comment by LoganWright** [comment]
I do want to comment on a couple things in the proposal that we can check off right away.

## Indexing

The using indexes for arrays is fully supported. I'm not fond of adding subscripting because it adds a new token which we should seek to avoid if it only duplicates an existing syntax. Right now, arrays are direct, like:

```
friends.0
```

It also has basic array value accessors

```
[
  "friends": [
        [
            ["name": "a"],
            ["name": "b"],
            ["name": "c"]
        ]
]
```

I can call:

```
#loop(friends.0.name, "name") { #(name) }
```

And I'll get:

```
a
b
c
```

This can be chained throughout

## Embed

This is there already, except we call it `include`. I'm open to a name change, but wanted to mention that the functionality is there 👍 

```
#include("includes/header")

Hello, other stuff, maybe body.

#include("includes/footer"
```

These will also be rendered within context.

**Comment by NathanFlurry** [comment]
My only concern with the `#format` function/commands/whatever is that it would require a variable number of arguments.

Another concept I forgot to put in the proposal is the idea of using pre-translated strings, e.g. someone could put `<p>strings[lang].somePhrase`. (This is the same for any templating engine, though.)

**Comment by tannernelson** [comment]
That's a really good point. We should find a way to include localization in this. 

Could look like:

```
<h1>#ln("header")</h1>
```

Or 

```
<h1>#(ln.header)</h1>
```

Where `ln` is some unique key. Could even be a symbol

**Comment by NathanFlurry** [comment]
@LoganWright The reason I included the subscript was because some people may want to index the data from a variable value. For instance, let's say (for some reason) someone wants to display every other friend, they would do this:

```
#loop(@i, 0, #count(friends.count)) {
	<p>friends[i].name</p>
}
```

However, one solution would be to do something like this instead: `#(friends.#(i).name)`.

**Comment by wildfiremedia** [comment]
Is parameter naming that Swift 3 uses relevant for template?

```
#loop(a: @i, b: 0, c: 10, d: 2, e: false) {
```

**Comment by BrettRToomey** [comment]
@wildfiremedia I don't know the exact answer, but my personal opinion is that we probably shouldn't have parameter naming. There are two main reasons I can think of instantly. The first being that we're trying to keep Leaf as simple/pretty as possible. The second is remembering tags without auto complete can be a bit of a pain and cause too much going back and forth between docs and your templates. On top of that, it would also add more work for the parser and slow down development for (what I personally see as) little gain.

**Comment by BrettRToomey** [comment]
@tannernelson @NathanFlurry Some stuff for localization is most definitely desired. Some type of unique global object, as Tanner suggested seems to make the most sense. It's more 🍃-esque and it keeps us from having to implement subscripts. 

A symbol could work nicely for this as we don't have to reserve keywords, but at the same time goes a little against Leaf's principles. Maybe we try reserving something like `strings` and maybe even reserving `locale` which would be a string representing the client's language code.

**Comment by LoganWright** [comment]
Hey everyone, I think localization stuff is slightly out of scope this early, tanner's `#ln` suggestion would fit most naturally right now, but let's first parse through what's here since there's so much already!

I'm gonna parse through all the headlines:

## Variables

Basically whether or not to include the `@`. I think a lot of the confusion associated w/ these distinctions is variable/constant don't really apply in a templating context. Let's try and get some good jargon around these two concepts so we have a better shared language:

- <#something-here#> = Key or KeyPath that will be pulled from context
- <#something-here#> = String declared inline that will be the same for every render

## Expressions

This is perhaps the most complex part of the proposal since it differs a lot from existing syntax, for example. `#(5 + 3)` requires separate variable parsing, and acknowledging spaces as well as what they mean in context, ie: `5 + 3` == `15`, not `"5" + "3" == "53"`. This is easier in things like Jade that actually just run Javascript. In the existing syntax, I think I'd add an `e` tag and parse manually, ie:

```
#e(age, ">", "21")
```

Would likely be:

```
#greaterThan(age, "21") { 
    <h1>Welcome to Bob's Beers</h1>
} ##else() {
    <h1>Must be 21 or older to view Bob's Beers</h1>
}
```

or

```
#add("5", "3")
```

I would say expressions are out of scope initially since they require a lot more knowledge about how parsing should go. They can also be created more manually, so we could provide some convenience tags similar to what's above.

## Functions

I think we're on same page here, everything listed should be possible currently 🎉 

## Indexing

Mentioned above, we provide full indexing, but you brought up a great point about dynamic indexes. In this situation, I think as opposed to adding a token initially, I'd opt for a new function, `index`. For example

```
Hello, #(friends[indexVariable])
```

Would be

```
Hello, #index(friends, indexVariable)
```

This is a common theme of us taking a bit of the long way. We really want to see what's possible in existing syntax rules so when we add things like subscripts, we can simply precompile them to expand long form w/o changing underlying render logic.

I think subscripts here would be great, but based on added complexity initially, I think I'd like to try the index stuff until it's no longer valid.

## Embedded Documents

As mentioned above, this is fully included, let me know if what's there doesn't work 👍 

## Blocks

Right now, the blocks are considered a "sub-tree" to some degree, in that any function can either have a sub-tree or not. If there is a subtree, the result of the function will be passed in at the top of the context's scope for rendering. We also have opted for a `##` (double token) to indicate chain. This allows any existing function to become a chain, for example, multiple if's can be changed w/o redefining an else-if

```
#if(a) {

} ##if(b) {

}
```

This also allows things like concatenating two distinct types

```
#greaterThan(age, "21") {
    <h1>Welcome to Bob's Beers</h1>
} ##include("includes/under-21-warning")
```

This says, if age from context is greater than static '21', show the header, otherwise, embed the leaf located at the given path.

These can also be nested out, for example, to get a prefix and uppercase it, I might do:

```
#prefix(firstName, "5") { #uppercased(self) }
```

This can theoretically continue ad infinitum. This can be clunky, and there will very likely be future sugar, but for now, the goal is to implement things in syntax and build on sugar later once we feel the underlying mechanics are stable.

## Data Types

We use Node which encompasses all listed data types 👍 

## Extending

#### Leaf Representable

Mentioned in DataTypes, but we're using Node, which means any NodeRepresentable object can be easily rendered. This can come straight from your Fluent database, a document type, or any other conforming type 🎉 

#### Functions

Functions are already fully extensible. In fact, 100% of leaf's functionality is added through extensions. There is no unique logic for default tags/functions. I think your implementation is considerably more straight forward and useful for basic use cases, so I added a `BasicTag` class that can override one simple function. Here's how `index` might look:

```swift
class Index: BasicTag {
    public init() {
        super.init(name: "index")
    }

    override func run(arguments: [Argument]) throws -> Node? {
        guard arguments.count == 2 else { return nil }
        guard let array = arguments[0].value?.nodeArray else { return nil }
        guard let index = arguments[1].value?.int, index < array.count else { return nil }
        return array[index]
    }
}
```

#### Template Extensions

Wasn't mentioned in your proposal, but wanted to talk about it. We definitely need this, and it's in the road map. Embedding is considerably easier than extensions since the inverse of embed is particularly tricky. This is definitely going to be in there, just requires some more work on my part.

# tl;dr

I think the biggest part up for discussion is better jargon around (pulled from context/static inline) parameter types and how we make a distinction between the two. `@`, `""`, `''`, `$`, etc.

Next is expressions and how we implement them consistently. My initial gut is that the syntax specified is beyond scope for this very early stage, but putting it here in case there's good solutions.

A lot of the manifesto stuff is no brainer, like easier leaf extensions and some mentioned functions. We'll get this stuff in straight away!

**Comment by BrettRToomey** [comment]
@LoganWright Because of so many years of seeing `""` as String literals, it was hard for my brain to get used to `""` being a variable and everything else (without a `#` prefix) being a literal. I think, initially, `""` is hard to read for those who are new to templates, but definitely makes sense once you start using it.

As for all the other prefix symbols, they look natural at first glance, but feels awkward once you start writing. Because, in other languages where you use `$` you have to use the symbol every time you interact with the variable, but in Leaf if you want to get the value of the variable you just enclose it with `#()`--making things even more confusing (imo). 

So, I think, it would be best to stay with `""` or maybe even something like `#var(friends)`. It's definitely explicit, but feels a little overkill.

**Comment by LoganWright** [comment]
There hasn't been much traffic here for a while. I think we'll gear up for 2.0 here in a few months after gathering more feedback. At that point, we'll re-engage. 

Thank you everyone for the lively and productive discussion 👍
